### PR TITLE
Data: remove the expando when there's no more data; minor cleanup

### PR DESCRIPTION
**Pull request by timmywil**
- This also removes the `Data#discard()` method, which we never used, and the `initial` paramater in `Data#register()` – also never used.

Fixes gh-1760

### CODE REVIEW

**Comment by rwaldron** [comment]
LGTM.

File: test/unit/data.js [hunk 1]
```diff
@@ -837,7 +837,24 @@ test("Check proper data removal of non-element descendants nodes (#8335)", 1, fu
...
 
-	equal(result, "ok", "enumeration of data- attrs on body" );
+	var key,
+		div = jQuery( "<div/>" );
+	div.data( "some", "data" );
+	equal( div.data( "some" ), "data", "Data is added" );
+	div.removeData( "some" );
+
+	// Make sure the expando is gone
+	for ( key in div[ 0 ] ) {
```
Comment (mzgol) [inline test/unit/data.js]: A similar test for on+off would be nice.

Comment (timmywil) [inline test/unit/data.js]: Sure, I'll add that.

Comment (timmywil) [inline test/unit/data.js]: Added

**Pull request closed without merging**